pkg/apis/core/types.go

```diff
@@ -2434,6 +2434,10 @@ const (
 	// disruption (such as preemption, eviction API or garbage-collection).
 	// The constant is to be renamed once the name is accepted within the KEP-3329.
 	AlphaNoCompatGuaranteeDisruptionTarget PodConditionType = "DisruptionTarget"
+	// ResourceExhausted indicates the pod is about to be deleted due to either
+	// exceeding its ephemeral storage limits or running an OOM killed container.
+	// The constant is to be renamed once the name is accepted within the KEP-3329.
+	AlphaNoCompatGuaranteeResourceExhausted = "ResourceExhausted"
```

Comment:
It is a new condition type. It is expected to be used as an indicator that a pod was terminated due to a non-retriable application or configuration issue. Example job failure policies using this condition are shown in the KEP:  
- https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#story-2
- https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#story-3

Reply:
OOM or ephemeral storage limits being reached doesn't necessarily mean the error is not retriable, right?

Reply:
>> is this condition expected to be mutually exclusive with DisruptionTarget?
>
> Not necessarily, it may happen in some race situations that both conditions are added (there is no mechanism to prevent). The user has control of giving priority to handling the condition types by the Job's podFailurePolicy rule order.

ok, I can see that if a container OOMed at the ~same time the low priority pod it was a part of was selected for eviction or something and independent actors both added their respective conditions.

maybe a better question is whether this condition is _orthogonal_ to DisruptionTarget, or if any of the code paths that add the `ResourceExhausted` condition would also add `DisruptionTarget`

Reply:
`ResourceExhausted` is only added by kubelet, so the search for non-orthogonal situations is restricted to kubelet.

I think there is one scenario in which they are non-ortoghonal - this is when the OS is low on memory. This situation can potentially result in both node memory pressure (resulting in `DisruptionTarget`) and OOM killing one of the pod containers (resulting in `ResourceExhausted`), even if the pod container's limits are not exceeded. 

This situation should be rare though. especially with the node configuration to leave enough memory available for the system so that OOMkiller would only happen if the pod container's limit is exceeded. It is further discussed in KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3329-retriable-and-non-retriable-failures#oom-killer-invoked-when-memory-limits-are-not-exceeded

Reply:
@liggitt @alculquicondor  do you have further concerns with the fields? 

There is also an example of a pod condition type `PodHasNetwork` used only in kubelet in alpha and is meant to be migrated later to API types: https://github.com/kubernetes/kubernetes/blob/6a6a36f5695029f142f165c58b41b2f585e383f9/pkg/kubelet/types/constants.go#L54. I guess I would need to either apply this strategy and keep `ResourceExhausted` there or declare it in the API. 

However, I prefer to declare it in the API early as this way we are consistent with `DisruptionTarget` and there is a better visibility that these two types of conditions are from the same KEP. Also, we solve the issue with potential naming change by prefixing it with `AlphaNoCompatGuarantee` to make it clear the name is subject to change while in alpha.

Reply:
> I think there is one scenario in which they are non-orthogonal - this is when the OS is low on memory. This situation can potentially result in both node memory pressure (resulting in DisruptionTarget) and OOM killing one of the pod containers (resulting in ResourceExhausted), even if the pod container's limits are not exceeded.

That's a great insight, and makes me wonder how a user can reliably write a rule to fail job pods on `ResourceExhausted` conditions if it *can* be due to things unrelated to the workload.

Reply:
The main thing I want to ensure is that any API surface is clear about how it should be used, and reliably works when used that way. Understanding whether this condition can be depended on to indicate there's a problem with the workload itself seems important

Reply:
Unfortunately, there is no reliable way to detect when a pod was OOMkilled due to memory being exhausted by other pods https://github.com/kubernetes/kubernetes/issues/112910

Users can have a higher confidence that ResourceExhausted is due to the pod itself if all their pods have memory requests==limits. And this is the use case we are trying to serve.

@mimowo can you include some of these details in the comment?

Reply:
Yes, the most important use case is serving workloads where all pods declare requests=limits for the containers. Admitting such pods indicate that the node has enough memory and when OOM killer is invoked users can interpret it as pod limits being exceeded. Still, technically some other process on the machine could cause the system to run out of memory and kill the workload container, but it seems an edge case.

When there are containers running with requests < limits, they could eat up the node's memory after being admitted and still not exceed the limits, but resulting in OOM killer. However, the invocation of OOM killer can generally be prevented by cluster configuration to leave enough memory to the system so that memory pressure is raised before OOM killer is invoked, following the practices here: https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#node-pressure-eviction-good-practices. When pods are evicted due to memory pressure DisruptionTarget condition is added.

Given the limitation of OOM killer not sure what can be done better except for documenting.

Reply:
FYI @liggitt it turns out that on CRI-O with cgroups v2 the OOM killed containers don't have the reason field set to `OOMKilled`, but generic `Error`. It seems like a bug in CRI-O, but discussion is open: https://github.com/kubernetes/kubernetes/pull/112977#issuecomment-1294980085.
We are yet discussing the importance of the feature for the use cases we have. Thus, considering to drop the resource exhausted from the KEP and only add DisruptionTarget condition in case of node pressure eviction + node graceful shutdown.